### PR TITLE
Add exclude option in selene.toml

### PR DESCRIPTION
**Pull request by kyojindev**
This implements an `exclude` glob option in selene.toml which excludes any files matching with the globs from being linted.

### CODE REVIEW

File: selene/Cargo.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ color-eyre = "0.6.1"
 dirs = "4.0.0"
 full_moon = "0.15.1"
 glob = "0.3"
+globset = "0.4.9"
```
Comment (Kampfkarren) [inline selene/Cargo.toml:23]: Can you replace our one usage of glob if you're going to add globset?

File: selene/src/main.rs [hunk 2]
```diff
@@ -469,6 +469,25 @@ fn start(mut matches: opts::Options) {
             }
         };
 
+    let mut builder = globset::GlobSetBuilder::new();
+    for pattern in &config.exclude {
+        builder.add(match globset::Glob::new(&format!("{}", pattern)) {
+            Ok(glob) => glob,
+            Err(error) => {
+                error!("Invalid glob pattern: {}", error);
```
Comment (Kampfkarren) [inline selene/src/main.rs:477]:
```suggestion
                error!("Invalid glob pattern: {error}");
```

File: selene/src/main.rs [hunk 3]
```diff
@@ -469,6 +469,25 @@ fn start(mut matches: opts::Options) {
             }
         };
 
+    let mut builder = globset::GlobSetBuilder::new();
+    for pattern in &config.exclude {
+        builder.add(match globset::Glob::new(&format!("{}", pattern)) {
```
Comment (Kampfkarren) [inline selene/src/main.rs:474]:
```suggestion
        builder.add(match globset::Glob::new(&pattern) {
```

File: selene/src/main.rs [hunk 4]
```diff
@@ -510,6 +529,9 @@ fn start(mut matches: opts::Options) {
                         for entry in glob {
                             match entry {
                                 Ok(path) => {
+                                    if exclude_set.is_match(&path) {
+                                        continue;
+                                    }
```
Comment (Kampfkarren) [inline selene/src/main.rs:534]:
```suggestion
                                    }

```

**Review by Kampfkarren — CHANGES_REQUESTED** [review]
Add it to CHANGELOG.md and to the documentation for selene.toml as well

File: selene/Cargo.toml (hunk 1 above)
Comment (kyojindev) [reply selene/Cargo.toml:23]: globset doesn't currently support iterating recursively through matching file paths, so I'd have to rewrite that entire section and do it manually. I can do it but I want to get the go-ahead from you before spending time on this.

Comment (Kampfkarren) [reply selene/Cargo.toml:23]: Hmm, so does that mean something like `a/**/b.lua` isn't possible to exclude? Or `a/**/*.spec.lua`? If glob behavior is different between glob and globset then we should just be using glob, and running over every glob itself. Maybe we should just be doing that anyway, I'm unsure what benefit globset has to just looping over a vector of globs.

Comment (kyojindev) [reply selene/Cargo.toml:23]: Sorry, I might have worded this wrong. Globset does allow those exclusions - what it doesn't allow is using a glob as an iterator like we currently do to recursively get the files to lint when provided with a directory.

Comment (Kampfkarren) [reply selene/Cargo.toml:23]: Works for me :+1:

**Comment by Kampfkarren** [comment]
Conflicts with master now

**Comment by Kampfkarren** [comment]
@OverHash Yeah, didn't catch it

**Pull request merged**